kafka replay speed: add alert for when we miss records in Kafka (#9921)
grafana/mimir: docs/sources/mimir/manage/mimir-runbooks/_index.md

  - If the call exists and it's waiting on a lock then there may be a deadlock.
  - If the call doesn't exist then it could either mean processing is not stuck (false positive) or the `pushToStorage` wasn't called at all, and so you should investigate the callers in the code.

### MimirIngesterMissedRecordsFromKafka

This alert fires when an ingester has missed processing some records from Kafka. In other words, there has been a gap in offsets.

How it **works**:

> - Ingester reads records from Kafka, and processes them sequentially. It keeps track of the offset of the last record it processed.

Reviewer: ```suggestion
- The ingester reads records from Kafka and processes them sequentially. It keeps track of the offset of the last record it's processed.
```

> - Upon fetching the next batch of records, it checks if the first available record has an offset one greater than the last processed offset. If the first available offset is larger than that, then the ingester has missed some records.

Reviewer: ```suggestion
- Upon fetching the next batch of records, it checks if the first available record has an offset of one greater than the last processed offset. If the first available offset is larger than that, then the ingester has missed some records.
```

> - Kafka doesn't guarantee sequential offsets. If a record has been manually deleted from Kafka or the records have been produced in a transaction and the transaction was aborted, then there may be a gap.

Reviewer: ```suggestion
- Kafka doesn't guarantee sequential offsets. If a record has been manually deleted from Kafka or if the records have been produced in a transaction and the transaction was aborted, then there may be a gap.
```

> - Mimir doesn't produce in transactions and does not delete records.

Reviewer: "Mimir doesn't produce in transactions" reads unclear to me. Is the "in" supposed to be here?

Author: yes. in Kafka you can create a transaction and in its context produce records.

> - When the ingester starts up, it will attempt to resume from the last offset it processed. If the ingester has been unavailable for long enough that the next record is already removed due to retention, then the ingester will miss some records.

Reviewer: ```suggestion
- When the ingester starts, it attempts to resume from the last offset it processed. If the ingester has been unavailable for long enough that the next record is already removed due to retention, then the ingester misses some records.
```

Reviewer: We avoid using future tense in the docs.

How to **investigate**:

> - Verify that there have been no deleted records in your Kafka cluster.

Author: ```suggestion
- Find the offsets which were missed. The ingester logs them along with the message `there is a gap in consumed offsets`.
- Verify that there have been no deleted records in your Kafka cluster.
```

> - Verify that there have been no deleted records in your Kafka cluster by humans or other applications.

Reviewer: I think we can probably remove "by humans or other applications".

Author: ```suggestion
- Verify that there have been no deleted records in your Kafka cluster.
```
